### ISSUE TITLE
Bi-directional text is unreliable

### ISSUE THREAD

**Issue by svenper** [reutenauer/polyglossia#204]
Loading the Arabic language and then not using it causes simple things like `\textenglish{Hello.}` to change the text direction. It can get even weirder if we use languages inside other language environments. XeLaTeX behaves as expected, but LuaLaTeX does not.

`\textenglish` and `\foreignlanguage{english}` have different strengths and weaknesses, sometimes using the right direction, sometimes breaking it, and sometimes showing nothing at all (if combined with `{otherlanguage}` or `{english}`.

```tex
% https://tex.stackexchange.com/questions/273301
\let\luatexpardir\pardir
\let\luatexlinedir\linedir
\let\luatextextdir\textdir
\let\luatexpagedir\textdir
\let\luatexbodydir\textdir

\documentclass{article}

\usepackage{fontspec}
\setmainfont{EB Garamond}
\newfontfamily\arabicfont{Scheherazade}[Script=Arabic,Scale=MatchUppercase]

\usepackage{polyglossia}
\setmainlanguage{english}
\setotherlanguage{arabic}
\setotherlanguage{french}

\begin{document}

\centering

\foreignlanguage{french}{Bonjour.}

Hello.

\foreignlanguage{arabic}{مرحبا.} Wrong direction.

\textarabic{مرحبا.}

Hello.

\textenglish{Hello.}

Hello.

\end{document}
```

![1](https://user-images.githubusercontent.com/15199675/38456686-a371995a-3a87-11e8-9cb5-1e817d339033.png)

**Comment by bastien-roucaries** [comment]
@svenper  Could you recheck with lastest git tree ?

**Comment by svenper** [comment]
@bastien-roucaries the same document compiles good now:

<img width="200" src="https://user-images.githubusercontent.com/15199675/63184754-9e25d000-c047-11e9-9cd6-df1aa02ea7bf.png" />

**Comment by bastien-roucaries** [comment]
@svenper 

Could you check with paragraph, long text and so on. Normally it work. Writting a demo will be nice and we will add to polyglossia. Normally if whole package is arabic indent is at right, for english at left.

**Comment by svenper** [comment]
I hacked together two small demos, one for mainlanguage arabic and one for english. It also highlights some other issues I found (now featuring xetex!):

* luatex and xetex output digits <sup lang="ar">0123456789</sup> <sup lang="ar">٠١٢٣٤٥٦٧٨٩</sup> <sup lang="fa">۰۱۲۳۴۵۶۷۸۹</sup> in the opposite direction from each other, but I don't know which one is correct.
* luatex outputs arabic.tex in the wrong column order.
* xetex outputs nested languages in the right character order but wrong word order.

[arabic.tex](https://gist.githubusercontent.com/svenper/4ccb37136b622ca10f7fbdb71ec24e82/raw)
[english.tex](https://gist.githubusercontent.com/svenper/ce9a5131990682a28eb158006171b10b/raw)

**Comment by seloumi** [comment]
For digits the correct output is  with xetex (numbers are always  LTR)

**Comment by jspitz** [comment]
Commit ae9250e6984 introduced a regression which needs to be fixed before 1.45.

The following MWE produces wrong output for embedded LTR text in section headings:
```tex
% !TeX TS-program = xelatex
\documentclass{article}%
\usepackage{polyglossia}

\setdefaultlanguage{arabic}
\setotherlanguage{english}

\setmainfont{Linux Libertine O}
\newfontfamily\arabicfont[Script=Arabic]{Amiri}%

\begin{document}

\tableofcontents

\section{أعمال تأريخية \textenglish{(Calendar operations)}}

\end{document}
```

The output looks like this:
![image](https://user-images.githubusercontent.com/4538783/67507973-55664500-f690-11e9-87ea-652c186c468f.png)

The only solution I found is to add explicit `\LRE` or `\RLE` in the `\xpg@textlanguage`:
```diff
diff --git a/tex/polyglossia.sty b/tex/polyglossia.sty
index 194ef48..f7da48b 100644
--- a/tex/polyglossia.sty
+++ b/tex/polyglossia.sty
@@ -154,6 +154,14 @@
      {}%
      {\xpg@error{Unknown~ language~ direction~ #1 ~(base~ wrapper)}}%
 }
+\cs_new_nopar:Nn{\polyglossia@settextdirection:n}{%
+  \str_case_e:nnTF{#1}{%
+       {LR}{\relax}%
+       {RL}{\xpg@error{right-to-left,~ but~ (lua)bidi~ package~ was~ not~ loaded!}}%
+     }%
+     {}%
+     {\xpg@error{Unknown~ language~ direction~ #1 ~(base~ wrapper)}}%
+}
 \AtEndOfPackageFile*{bidi}{%
   \ExplSyntaxOn%
   \cs_gset_nopar:Nn{\polyglossia@setdirection:n}{%
@@ -164,6 +172,14 @@
       {}%
       {\xpg@error{Unknown~ language~ direction~ #1 ~(bidi~ wrapper)}}%
   }%
+  \cs_gset_nopar:Nn{\polyglossia@settextdirection:n}{%
+    \str_case_e:nnTF{#1}{%
+        {LR}{\LRE}%
+        {RL}{\RLE}%
+      }%
+      {}%
+      {\xpg@error{Unknown~ language~ direction~ #1 ~(bidi~ wrapper)}}%
+  }%
   \ExplSyntaxOff%
 }
 \AtEndOfPackageFile*{luabidi}{%
@@ -176,6 +192,14 @@
       {}%
       {\xpg@error{Unknown~ language~ direction~ #1 ~(luabidi~ wrapper)}}%
   }%
+  \cs_gset_nopar:Nn{\polyglossia@settextdirection:n}{%
+    \str_case_e:nnTF{#1}{%
+        {LR}{\LRE}%
+        {RL}{\RLE}%
+      }
+      {}%
+      {\xpg@error{Unknown~ language~ direction~ #1 ~(luabidi~ wrapper)}}%
+  }%
   \ExplSyntaxOff%
 }
 
@@ -682,6 +706,18 @@
       }
 }
 
+
+\cs_new:Nn{\polyglossia@lang@settextdirection:nn}{
+  \prop_get:NnNTF \polyglossia@langsetup {#1/direction} \l_tmpa_tl
+      {
+        \polyglossia@settextdirection:n{\l_tmpa_tl}{#2}
+      }
+      {
+        \xpg@error{Could~ not~ retrieve~ key~ direction~ for~ language~ "#1"}
+        \prop_show:N{\polyglossia@langsetup}
+      }
+}
+
 \AtEndDocument{\prop_log:N{\polyglossia@langsetup}}
 \def\xpg@lastlanguage{0}%
 
@@ -1247,7 +1283,7 @@
    % This is a hook for external packages which want to access variants
    % via babelname (such as biblatex)
    \cs_if_exist_use:c{date@bbl@\babelname}%
-   #3
+   \polyglossia@lang@settextdirection:nn{#2}{#3}
    \egroup
   % Reset the language's/script's font families
   \str_if_eq:eeTF{\prop_item:Nn{\polyglossia@langsetup}{#2/lcscript}}{latin}{}{\resetfontlatin}%
```

@bastien-roucaries, @reutenauer, please have a look.

**Comment by jspitz** [comment]
The original bug is also partially reproducable with XeTeX. It seems I removed some `\leavevmode`s too much in f2f09c38c8dc78

**Comment by jspitz** [comment]
In the recent update of #329 the issue is fixed properly. I have separated bidi commands to change paragraph direction (`\setRL` etc.) from those which change inline text direction (`\RLE` etc.). Now the inline language changes (`\text<lang>`, `\foreignlanguage`, `otherlanguage*`) use the inline text direction switches, the `<lang>` environment, `\selectlanguage` and `otherlanguage` use the paragraph direction switches. This fixes all bugs reported here, and the `\leavevmode` trick (which basically just disables the paragraph direction switches in inline commands) is no longer needed.

@reutenauer can I merge this in as a last-minute fix for. 1.45?

**Comment by jspitz** [comment]
> I hacked together two small demos, one for mainlanguage arabic and one for english. It also highlights some other issues I found (now featuring xetex!):
> 
>     * luatex and xetex output digits 0123456789 ٠١٢٣٤٥٦٧٨٩ ۰۱۲۳۴۵۶۷۸۹ in the opposite direction from each other, but I don't know which one is correct.
> 
>     * luatex outputs arabic.tex in the wrong column order.

These differences are due to deficiencies of `luabidi` (as opposed to `bidi`). The latter automatically re-orders columns and handles numbers (and does many other thing `luabidi` doesn't). You might file an enhancement request at https://github.com/bidi-tex/luabidi, but there is nothing we can do here on the polyglossia side.

>     * xetex outputs nested languages in the right character order but wrong word order.

This is fixed by #329

**Comment by jspitz** [comment]
I merged it now, since it really should be fixed before 1.45.

I'm holding back everything else for after the release.
